packages/@aws-cdk/example-construct-library/lib/example-resource.ts

```diff
@@ -0,0 +1,501 @@
+ *  in our example, those are {@link exampleResourceArn} and {@link exampleResourceName}.
+ *
+ *  The interface defines the behaviors the resource exhibits.
+ *  Common behaviors are:
+ *    - {@link addToRolePolicy} for resources that are tied to an IAM Role
+ *    - grantXyz() methods (represented by {@link grantRead} in this example)
+ *    - onXyz() CloudWatch Events methods (represented by {@link onEvent} in this example)
+ *    - metricXyz() CloudWatch Metric methods (represented by {@link metricSth} in this example)
+ *
+ *  Of course, other behaviors are possible -
+ *  it all depends on the capabilities of the underlying resource that is being modeled.
+ *
+ *  This interface must always extend the IResource interface from the core module.
+ *  It can also extend some other common interfaces that add various default behaviors -
+ *  some examples are shown below.
+ */
+export interface IExampleResource extends
+  // all L2 interfaces need to extend IResource
+  core.IResource,
+
+  // Only for resources that have an associated IAM Role.
+  // Allows this resource to be the target in calls like bucket.grantRead(exampleResource).
+  iam.IGrantable,
+
+  // only for resources that are in a VPC and have SecurityGroups controlling their traffic
+  ec2.IConnectable {
+
+  /**
+   * The ARN of example resource.
+   * Equivalent to doing `{ 'Fn::GetAtt': ['LogicalId', 'Arn' ]}`
+   * in CloudFormation.
+   * It needs to be annotated with '@attribute'.
+   *
+   * @attribute
+   */
+  readonly exampleResourceArn: string;
+
+  /**
+   * The physical name of the example resource.
+   * Often, equivalent to doing `{ 'Ref': 'LogicalId' }`
+   * (but not always - depends on the particular resource modeled)
+   * in CloudFormation.
+   * Also needs to be annotated with '@attribute'.
+   *
+   * @attribute
+   */
+  readonly exampleResourceName: string;
+
+  /**
+   * For resources that have an associated IAM Role,
+   * surface that Role as a property,
+   * so that other classes can add permissions to it.
+   * Make it optional,
+   * as resources imported with {@link ExampleResource.fromExampleResourceName}
+   * will not have this set.
+   */
+  readonly role?: iam.IRole;
+
+  /**
+   * For resources that have an associated IAM Role,
+   * surface a method that allows you to conditionally
+   * add a statement to that Role if it's known.
+   * This is just a convenience,
+   * so that clients of your interface don't have to check {@link role} for null.
+   * Many such methods in the CDK return void;
+   * you can also return a boolean indicating whether the permissions were in fact added
+   * (so, when {@link role} is not null).
+   */
+  addToRolePolicy(policyStatement: iam.PolicyStatement): boolean;
+
+  /**
+   * An example of a method that grants the given IAM identity
+   * permissions to this resource
+   * (in this case - read permissions).
+   */
+  grantRead(identity: iam.IGrantable): iam.Grant;
```

Comment:
Shouldn't also have a `grant()` method which can take in a list of actions?

Reply:
`grant` isn't always required, and I'm not sure if it adds much to this example beyond `grantRead` (I'm trying to keep the example as small as possible, and even with that it's grown pretty big).

Reply:
I'm fine with this as is.